raft/multinode.go

```diff
@@ -143,24 +143,23 @@ func (g *groupState) commitReady(rd Ready) {
 	if !IsEmptyHardState(rd.HardState) {
 		g.prevHardSt = rd.HardState
 	}
-	if !IsEmptySnap(rd.Snapshot) {
-		g.prevSnapi = rd.Snapshot.Metadata.Index
-		g.raft.raftLog.stableSnapTo(g.prevSnapi)
+	if g.prevHardSt.Commit != 0 {
+		// In most cases, prevHardSt and rd.HardState will be the same because
```

Comment:
No, I mean prevHardSt. The old code used CommittedEntries (and had rd.HardState commented out), but neither is exactly correct. The difference is that when the HardState is restored from storage, it is immediately set as prevHardSt but it is not sent out as rd.HardState (since it is already persisted). When restoring from a snapshot with an empty log, we must use prevHardSt to set the applied index, or we may try to re-apply entries that are already incorporated into the snapshot. (This can occur if applying a snapshot and truncating the log are separate operations; I haven't been able to construct a test for this because in MemoryStorage the log is always truncated by applying a snapshot. This will generally result in a panic because this node will not have all of the entries that precede the snapshot).